pallets/subspace/src/math.rs

```diff
@@ -438,576 +39,102 @@ pub fn mask_diag_sparse(sparse_matrix: &Vec<Vec<(u16, I32F32)>>) -> Vec<Vec<(u16
-            result[i] += vector[j] * matrix[i][j];
-        }
-    }
-    result
-}
 
-// Row-wise sparse_matrix-vector product, column-wise sum: result_j = SUM(i) vector_i * matrix_ij.
-#[allow(dead_code)]
-pub fn matmul_sparse(
-    sparse_matrix: &Vec<Vec<(u16, I32F32)>>,
-    vector: &Vec<I32F32>,
-    n: u16,
-) -> Vec<I32F32> {
-    let mut result: Vec<I32F32> = vec![I32F32::from_num(0.0); n as usize];
-    for (i, sparse_row) in sparse_matrix.iter().enumerate() {
-        for (j, value) in sparse_row.iter() {
-            // Compute trust scores: t_j = SUM(i) w_ij * s_i
-            // result_j = SUM(i) vector_i * matrix_ij
-            result[*j as usize] += vector[i] * value;
+    /// Reshape vector to sparse matrix with specified number of input rows, cast f32 to I32F32.
+    fn vec_to_sparse_mat_fixed(
+        vector: &[f32],
+        rows: usize,
+        transpose: bool,
+    ) -> Vec<Vec<(u16, I32F32)>> {
+        assert!(
+            vector.len() % rows == 0,
+            "Vector of len {:?} cannot reshape to {rows} rows.",
+            vector.len()
+        );
+        let cols: usize = vector.len() / rows;
+        let mut mat: Vec<Vec<(u16, I32F32)>> = vec![];
+        if transpose {
+            for col in 0..cols {
+                let mut row_vec: Vec<(u16, I32F32)> = vec![];
+                for row in 0..rows {
+                    if vector[row * cols + col] > 0. {
+                        row_vec.push((row as u16, I32F32::from_num(vector[row * cols + col])));
+                    }
+                }
+                mat.push(row_vec);
+            }
+        } else {
+            for row in 0..rows {
+                let mut row_vec: Vec<(u16, I32F32)> = vec![];
+                for col in 0..cols {
+                    if vector[row * cols + col] > 0. {
+                        row_vec.push((col as u16, I32F32::from_num(vector[row * cols + col])));
+                    }
+                }
+                mat.push(row_vec);
+            }
         }
+        mat
     }
-    result
-}
 
-// Column-wise sparse_matrix-vector product, row-wise sum: result_i = SUM(j) vector_j * matrix_ij.
-#[allow(dead_code)]
-pub fn matmul_transpose_sparse(
-    sparse_matrix: &Vec<Vec<(u16, I32F32)>>,
-    vector: &Vec<I32F32>,
-) -> Vec<I32F32> {
-    let mut result: Vec<I32F32> = vec![I32F32::from_num(0.0); vector.len()];
-    for (i, sparse_row) in sparse_matrix.iter().enumerate() {
-        for (j, value) in sparse_row.iter() {
-            // Compute dividends: d_j = SUM(i) b_ji * inc_i
-            // result_j = SUM(i) vector_i * matrix_ji
-            // result_i = SUM(j) vector_j * matrix_ij
-            result[i] += vector[*j as usize] * value;
+    /// Returns a normalized (sum to 1 except 0) copy of the input vector.
+    fn normalize(x: &[I32F32]) -> Vec<I32F32> {
+        let x_sum: I32F32 = x.iter().sum();
+        if x_sum == I32F32::from_num(0.0) {
+            x.into()
```

Comment:
O `to_vec` é mais claro na comunicação
```suggestion
            x.to_vec()
```